core/modules/drr.cc

```diff
@@ -0,0 +1,392 @@
+  if (err_num != 0) {
+    return pb_errno(err_num);
+  }
+  current_flow_ = nullptr;
+  return pb_errno(0);
+}
+
+void DRR::DeInit() {
+  for (auto it=flows_.begin(); it != flows_.end(); ) {
+    RemoveFlow(it->second);
+    it++;
+  }
+  mem_free(flow_ring_);
+}
+
+pb_cmd_response_t DRR::CommandQuantumSize(const bess::pb::DRRQuantumArg &arg) {
+  pb_cmd_response_t response;
+  set_cmd_response_error(&response, SetQuantumSize(arg.quantum()));
+  return response;
+}
+
+pb_cmd_response_t DRR::CommandMaxFlowQueueSize(
+      const bess::pb::DRRMaxFlowQueueSizeArg &arg) {
+  pb_cmd_response_t response;
+  set_cmd_response_error(&response, SetMaxFlowQueueSize(arg.max_queue_size()));
+  return response;
+}
+
+void DRR::ProcessBatch(bess::PacketBatch *batch) {
+  int err;
+
+  // insert packets in the batch into their corresponding flows
+  for (int i = 0; i < batch->cnt(); i++) {
+    bess::Packet *pkt = batch->pkts()[i];
+
+    // TODO(joshua): Add support for fragmented packets.
+    Flow_Id id = GetId(pkt);
+    auto it = flows_.Find(id);
+
+    // if the Flow doesn't exist create one
+    // and add the packet to the new Flow
+    if (it == nullptr) {
+      if (llring_full(flow_ring_)) {
+        bess::Packet::Free(pkt);
+      } else {
+        err = AddNewFlow(pkt, id);
+        assert(err == 0);
+      }
+    } else {
+      err = Enqueue(it->second, pkt);
+      assert(err == 0);
+    }
+  }
+  err++;
+}
+
+struct task_result DRR::RunTask(void *) {
+  bess::PacketBatch batch;
+  struct task_result ret;
+
+  batch.clear();
+  int total_bytes = 0;
+  if (flow_ring_ != NULL) {
+     total_bytes = GetNextBatch(&batch);
+  }
+  assert(total_bytes >= 0);  // TODO(joshua) do proper error checking
+  total_bytes++;
+
+  uint64_t cnt = batch.cnt();
+  if (cnt > 0) {
+    RunNextModule(&batch);
+  }
+
+  ret = (struct task_result){
+      .packets = cnt, .bits = (total_bytes + cnt * kPACKET_OVERHEAD) * 8,
+  };
+
+  return ret;
+}
+
+int DRR::GetNextBatch(bess::PacketBatch* batch) {
+  int err;
+  double now = get_epoch_time();
+  Flow* f;
+  uint32_t total_bytes = 0;
+  int count = llring_count(flow_ring_);
+  int batch_size = batch->cnt();
+
+  // iterate through flows in round robin fashion until batch is full
+  while (!batch->full()) {
+    // checks to see if there has been no update after a full round
+    if (count == 0) {
+      if (batch_size == batch->cnt()) {
+        break;
+      } else {
+        count = llring_count(flow_ring_);
+        batch_size = batch->cnt();
+      }
+    }
+    count--;
+
+    // gets the next flow in the round
+    if (!current_flow_) {
+      err =  llring_sc_dequeue(flow_ring_, reinterpret_cast<void**>(&f));
+      if (err < 0)
+        return err;
+
+      if (llring_empty(f->queue) && !f->next_packet) {
+        // if the flow expired, remove it
+        if (now - f->timer > kTTL) {
+          RemoveFlow(f);
+        } else {
+          err = llring_sp_enqueue(flow_ring_, f);
```

Comment:
Assuming its 187. there is a special case where you have to check if the non-empty queue becomes empty after dequeuing its packets so that you can reset its deficit counter to zero. However, if the queue has been empty, you don't reset its deficit counter. so, you check before trying to dequeue if it is an empty flow, and if so, you renqueue and continue in the while loop.